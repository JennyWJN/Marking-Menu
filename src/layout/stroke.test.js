import stroke from './stroke';

const docCreateElement = document.createElement;

const createMockContext = () =>
  new Proxy(
    {
      mock: {
        methodCalls: []
      }
    },
    {
      get(target, name) {
        return name in target
          ? target[name]
          : (...args) => {
              target.mock.methodCalls.push({
                method: name,
                args
              });
            };
      }
    }
  );

beforeEach(() => {
  document.createElement = jest.fn((t, ...args) => {
    const elt = docCreateElement.call(document, t, ...args);
    if (t === 'canvas') {
      const context = createMockContext();
      elt.getContext = () => context;
    }
    return elt;
  });
});

afterEach(() => {
  document.createElement = docCreateElement;
});

describe('stroke', () => {
  it('creates its DOM and set up the canvas scale', () => {
    const div = document.createElement('div');
    div.getBoundingClientRect = () => ({ width: 50, height: 60 });
    stroke(div, { doc: document, ptSize: 1 / 4 });
    expect(div).toMatchSnapshot();
    expect(div.querySelector('canvas').getContext().mock.methodCalls).toEqual([
      { method: 'scale', args: [4, 4] }
    ]);
  });

  it('can be removed', () => {
    const div = document.createElement('div');
    div.getBoundingClientRect = () => ({ width: 50, height: 60 });
    const strokeCanvas = stroke(div, { doc: document, ptSize: 1 / 4 });
    strokeCanvas.remove();
    expect(div).toMatchSnapshot();
  });

  it('can draw points', () => {
    const div = document.createElement('div');
    div.getBoundingClientRect = () => ({ width: 50, height: 60 });
    const strokeCanvas = stroke(div, {
      doc: document,
      ptSize: 1 / 4,
      pointRadius: 100,
      pointColor: 'mockPointColor'
    });
    const mockContext = div.querySelector('canvas').getContext();
    // Clear the method calls.
    mockContext.mock.methodCalls = [];

    strokeCanvas.drawPoint([10, 12]);
    strokeCanvas.drawPoint([5.2, 0]);
    expect(mockContext.mock.methodCalls).toMatchSnapshot();
  });

  it('can draw strokes', () => {
    const div = document.createElement('div');
    div.getBoundingClientRect = () => ({ width: 50, height: 60 });
    const strokeCanvas = stroke(div, {
      doc: document,
      ptSize: 1 / 4,
      lineWidth: 7,
      lineColor: 'mockLineColor'
    });
    const mockContext = div.querySelector('canvas').getContext();
    // Clear the method calls.
    mockContext.mock.methodCalls = [];

    strokeCanvas.drawStroke([[3, 2], [2, 4], [5, 1], [9, 0]]);
    strokeCanvas.drawStroke([[10, 0], [0, 0], [1, 1], [1, 0]]);
    expect(mockContext.mock.methodCalls).toMatchSnapshot();
  });
});
