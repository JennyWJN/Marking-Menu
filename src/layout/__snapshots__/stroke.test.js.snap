// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stroke can be removed 1`] = `<div />`;

exports[`stroke can draw points 1`] = `
Array [
  Object {
    "args": Array [],
    "method": "save",
  },
  Object {
    "args": Array [],
    "method": "beginPath",
  },
  Object {
    "args": Array [
      110,
      12,
    ],
    "method": "moveTo",
  },
  Object {
    "args": Array [
      10,
      12,
      100,
      0,
      360,
    ],
    "method": "arc",
  },
  Object {
    "args": Array [],
    "method": "fill",
  },
  Object {
    "args": Array [],
    "method": "restore",
  },
  Object {
    "args": Array [],
    "method": "save",
  },
  Object {
    "args": Array [],
    "method": "beginPath",
  },
  Object {
    "args": Array [
      105.2,
      0,
    ],
    "method": "moveTo",
  },
  Object {
    "args": Array [
      5.2,
      0,
      100,
      0,
      360,
    ],
    "method": "arc",
  },
  Object {
    "args": Array [],
    "method": "fill",
  },
  Object {
    "args": Array [],
    "method": "restore",
  },
]
`;

exports[`stroke can draw strokes 1`] = `
Array [
  Object {
    "args": Array [],
    "method": "save",
  },
  Object {
    "args": Array [],
    "method": "beginPath",
  },
  Object {
    "args": Array [
      3,
      2,
    ],
    "method": "moveTo",
  },
  Object {
    "args": Array [
      2,
      4,
    ],
    "method": "lineTo",
  },
  Object {
    "args": Array [
      5,
      1,
    ],
    "method": "lineTo",
  },
  Object {
    "args": Array [
      9,
      0,
    ],
    "method": "lineTo",
  },
  Object {
    "args": Array [],
    "method": "stroke",
  },
  Object {
    "args": Array [],
    "method": "restore",
  },
  Object {
    "args": Array [],
    "method": "save",
  },
  Object {
    "args": Array [],
    "method": "beginPath",
  },
  Object {
    "args": Array [
      10,
      0,
    ],
    "method": "moveTo",
  },
  Object {
    "args": Array [
      0,
      0,
    ],
    "method": "lineTo",
  },
  Object {
    "args": Array [
      1,
      1,
    ],
    "method": "lineTo",
  },
  Object {
    "args": Array [
      1,
      0,
    ],
    "method": "lineTo",
  },
  Object {
    "args": Array [],
    "method": "stroke",
  },
  Object {
    "args": Array [],
    "method": "restore",
  },
]
`;

exports[`stroke creates its DOM and set up the canvas scale 1`] = `
<div>
  <canvas
    height="240"
    style="position: absolute; left: 0px; top: 0px; width: 50px; height: 60px; pointer-events: none;"
    width="200"
  />
</div>
`;
